packages/automerge-repo-react-hooks/src/useBootstrap.ts

```diff
@@ -0,0 +1,58 @@
+import { useRepo } from "automerge-repo-react-hooks";
+import { DocumentId } from "automerge-repo";
+import { useEffect, useState } from "react";
+
+// Set URL hash
+export const setHash = (hash: string, pushState = false) =>
+  void history[pushState ? "pushState" : "replaceState"]("", "", "#" + hash);
+
+// Get current URL hash
+export const useHash = () => {
+  const [hashValue, setHashValue] = useState(window.location.hash);
+  useEffect(() => {
+    const handler = () => void setHashValue(window.location.hash);
+    window.addEventListener("hashchange", handler);
+    return () => void window.removeEventListener("hashchange", handler);
+  }, []);
+  return hashValue;
+};
+
+// Get a key from a query-param-style hash URL
+export const getQueryParamValue = (key: string, hash) =>
+  hash.match(new RegExp(`${key}=([^&]*)`))?.[1];
+
+export const createDocument = (repo, onCreate) => {
+  const handle = repo.create(); // Create a new document
+  if (onCreate) handle.change(onCreate); // Set initial state
+  return handle;
+};
+
+export const useBootstrap = ({
+  onCreate = () => {},
+  hashRouteKey = "id",
+  localStorageKey = "documentId",
+  getDocumentId = (hash) =>
+    getQueryParamValue(hashRouteKey, hash) ??
+    (localStorageKey && localStorage.getItem(localStorageKey)),
+  setDocumentId = (documentId) => {
+    // Only set hashroute if document ID changed
+    if (documentId !== getQueryParamValue(hashRouteKey, window.location.hash))
+      setHash(`${hashRouteKey}=${documentId}`);
+    if (localStorageKey) localStorage.setItem(localStorageKey, documentId);
+  },
+} = {}) => {
+  const repo = useRepo();
+  const hash = useHash();
+
+  const existingDocumentId = getDocumentId(hash) as DocumentId | undefined;
+
+  const handle = existingDocumentId
+    ? repo.find(existingDocumentId) // TODO: Handle bad existingDocumentId

```

Comment:
Not sure how to go about handling a non-existing document, if you give it a bad ID the document just never becomes ready

Reply:
presumably the app (or a separate hook) should watch `repo.on("document", ({handle: DocHandle}) => void)`, the app UI can be responsible for showing an loading message and presenting options if it never loads

Reply:
I'm going to update the ID-creation code to use checksummed bs58 IDs (or something like that.) This will allow us to differentiate between transcription errors or unavailable data (which ultimately cannot be prevented.)

Reply:
Sounds good, if the document is still not ready after 30 seconds I can show a screen saying "can't seem to load document, keep trying? or try something else?"